### PR TITLE
Add endpoints, swagger and Docs

### PR DESCRIPTION
**Pull request by CharlieGarcia**
- Add some API endpoints like `/api/v1/products`, `/api/v1/products/:id`, `/api/v1/docs`.
- Add swagger for a better API explanation and documentation.
- Add development documentation to the project.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -1,2 +1,52 @@
-# agora
-Agora
+# Agora
+This the project API for products in NY Magazine.
```
Comment (Shilario) [inline README.md]:
```suggestion
This is the project API for products in NY Magazine.
```

File: README.md [hunk 2]
```diff
@@ -1,2 +1,52 @@
...
+- Make.
+- A [Postgres](https://www.postgresql.org/) database instance to connect.
+
+## Setting up the project
+- Open the terminal and `cd` to the project root directory.
+- Type `make` and press the `Return` button.
+
+That should install all the project dependencies and initialize the database instance and the products tables with it.
+
+In the sake of simplicity you can visit the [Adminer instance](http://localhost:9080) for this project and verify the existence of the database, the tables structure and even run some SQL commands there.
```
Comment (manuelurenah) [inline README.md]: Small suggestion
```suggestion
For the sake of simplicity you can visit the [Adminer instance](http://localhost:9080) for this project and verify the existence of the database, the tables structure and even run some SQL commands there.
```

File: README.md [hunk 3]
```diff
@@ -1,2 +1,52 @@
...
+
+Followed by
+```
+psql -U postgres
+```
+And voilà you are inside the Postgres container and using psql.
+
+## Endpoints
+
+This API counts with some endpoints from the base url which is http://localhost:9000/api/v1 to interact with between them we have:
```
Comment (manuelurenah) [inline README.md]: Small rephrasing
```suggestion
This API has a series of endpoints from the base url which is http://localhost:9000/api/v1 to interact with between them we have:
```

File: api/v1/router/index.js [hunk 4]
```diff
@@ -0,0 +1,31 @@
...
+    .then(data => data.length ? resp.status(200).json(data) : resp.status(404).json([]))
+    .catch(error => resp.status(500).json({ statusCode: 500, message: error }));
+});
+
+router.get('/product/:id', (req, resp) => {
+  const { id } = req.params;
+
+  return controller.getProductById(id)
+    .then(data => data.length ? resp.status(200).json(data) : resp.status(404).json([]))
+    .catch(error => resp.status(500).json({ statusCode: 500, message: error }));
```
Comment (manuelurenah) [inline api/v1/router/index.js]: We could have a `handleError` method and reuse it in every endpoint

File: db/client.js [hunk 5]
```diff
@@ -0,0 +1,75 @@
+'use strict';
+
+const knexConnection = require('knex'),
+  { PG_USER = 'postgres', PG_PWD = 'example', DB_HOST = 'postgres', DB_NAME = 'products_db' } = process.env;
```
Comment (manuelurenah) [inline db/client.js]: Hmmm, what do you think if we set these constants properties inside the `constants.js` file?

File: db/client.js [hunk 6]
```diff
@@ -0,0 +1,75 @@
...
+}
+
+/**
+ * Gets products from database
+ * @param [_limit] query limit - default 10
+ * @returns {Promise}
+ */
+function getProducts(_limit) {
+  return knex.with('product_agg',
+    knex.raw('SELECT p.*, json_agg(v.*) vendors FROM products p LEFT JOIN vendors v ON v.product_id = p.product_id GROUP BY 1'))
```
Comment (Shilario) [inline db/client.js]: Maybe we should consider to remove the alias of the tables, the names aren't that long and is kind of hard to follow what they mean because they are not meaningful.

File: swagger.json [hunk 7]
```diff
@@ -0,0 +1,197 @@
...
+            "in": "query",
+            "description": "Limit the all products query result",
+            "required": false,
+            "type": "number",
+            "default": 10
+          }
+        ],
+        "responses": {
+          "200": {
+            "description": "successful operation",
```
Comment (Shilario) [inline swagger.json]:
```suggestion
            "description": "Successful operation",
```

File: swagger.json [hunk 8]
```diff
@@ -0,0 +1,197 @@
...
+          "404": {
+            "description": "Product not found"
+          },
+          "405": {
+            "description": "Validation exception"
+          }
+        }
+      }
+    },
+    "/product/id": {
```
Comment (Shilario) [inline swagger.json]: I think this way it will let you specified a param in the url.
```suggestion
    "/product/{id}": {
```

File: swagger.json [hunk 9]
```diff
@@ -0,0 +1,197 @@
...
+    }
+  },
+  "securityDefinitions": {
+    "petstore_auth": {
+      "type": "oauth2",
+      "authorizationUrl": "https://petstore.swagger.io/oauth/authorize",
+      "flow": "implicit",
+      "scopes": {
+        "write:pets": "modify pets in your account",
+        "read:pets": "read your pets"
```
Comment (reubenson) [inline swagger.json]: 🐕

File: scripts/postgres/init.sql [hunk 10]
```diff
@@ -34,13 +20,6 @@ last_modified TIMESTAMPTZ DEFAULT now(),
...
 
-CREATE TABLE IF NOT EXISTS page_uris (
-id TEXT DEFAULT uuid_generate_v1mc(),
-product_id TEXT REFERENCES products,
-uri TEXT,
-PRIMARY KEY(id)
-);
-
 CREATE TABLE IF NOT EXISTS vendors (
 id TEXT DEFAULT uuid_generate_v1mc(),
```
Comment (reubenson) [inline scripts/postgres/init.sql]: we can follow up in a separate ticket, but product id has a character length limit (historically, 6)

File: README.md [hunk 11]
```diff
@@ -1,2 +1,52 @@
...
+  - *POST*:
+    Create a new product in the database.
+  
+- /product/:id
+  - *GET*:
+    Display a product based in the product id.
+
+- /docs
+  - *GET*:
+    Display a swagger server displaying more information about the API endpoints and also allows to test the API endpoints.
```
Comment (reubenson) [inline README.md]: 🙌

File: api/v1/router/index.js [hunk 12]
```diff
@@ -0,0 +1,31 @@
...
+});
+
```
Comment (reubenson) [inline api/v1/router/index.js]: missing router to handle `/docs`?

File: db/client.js [hunk 13]
```diff
@@ -0,0 +1,75 @@
...
+ * @param {Object} productObj 
+ * @return {Promise}
+ */
+function addProduct(productObj) {
+  const { name, upc, description, image_url } = productObj;
+
+  return knex.transaction((trx) =>
+    trx.returning('product_id').insert({ name, upc, description, image_url }).into('products')
+      .then((product_id) => addVendors(trx, Object.assign(productObj, { product_id: product_id[0] }))));
+}
```
Comment (reubenson) [inline db/client.js]: it probably makes sense to follow up this PR with another ticket that specifies more granular detail like validation, but just noting here that going forward we'll want this operation to fail if name is not provided
